Diamond inheritance incorrectly detected as circular

Diamond inheritance patterns are incorrectly flagged as circular inheritance. This is a valid Python pattern that uses the Method Resolution Order (MRO) to handle shared base classes.

Reproduction:

```python
# base.py
class BaseTest:
    pass

class BaseA(BaseTest):
    pass

class BaseB(BaseTest):
    pass

class DerivedTest(BaseA, BaseB):  # Diamond inheritance
    pass
```

```python
# test_example.py
from base import DerivedTest

class TestExample(DerivedTest):
    def test_something(self):
        assert True
```

Running `rtest --collect-only test_example.py` produces:

```
ERROR collecting test_example.py
E   Circular inheritance detected in class hierarchy involving 'BaseTest'
```

The same file collects fine with pytest. Diamond inheritance is valid - Python's MRO handles this correctly. The cycle detection should track visited classes per inheritance path rather than globally, so that a class appearing in multiple branches isn't treated as a cycle.

--- a/src/python_discovery/semantic_analyzer.rs
+++ b/src/python_discovery/semantic_analyzer.rs
@@ -534,21 +534,29 @@ impl SemanticTestDiscovery {
         // Get the base class info
         let base_info = match self.class_cache.get(&key) {
             Some(info) => info.clone(),
-            None => return Ok(false),
+            None => {
+                visited.remove(&key);
+                return Ok(false);
+            }
         };
 
         // If this class has __init__, return true
         if base_info.has_init {
+            visited.remove(&key);
             return Ok(true);
         }
 
         // Otherwise, recursively check all parent classes
         for parent in &base_info.base_classes {
             if self.base_class_has_init_impl(parent, module_resolver, visited)? {
+                visited.remove(&key);
                 return Ok(true);
             }
         }
 
+        // Remove from visited after processing to allow sibling branches in diamond
+        // inheritance patterns to revisit this class without false cycle detection
+        visited.remove(&key);
         Ok(false)
     }
 
@@ -593,6 +601,9 @@ impl SemanticTestDiscovery {
         let base_info = match self.class_cache.get(&key) {
             Some(info) => info.clone(),
             None => {
+                // Remove from visited before early return to allow sibling branches
+                // to process this class without false cycle detection
+                visited.remove(&key);
                 return Ok(None);
             }
         };
@@ -610,6 +621,10 @@ impl SemanticTestDiscovery {
 
         // Then add this class's own methods
         all_methods.extend(base_info.test_methods.clone());
+
+        // Remove from visited after processing to allow sibling branches in diamond
+        // inheritance patterns to revisit this class without false cycle detection
+        visited.remove(&key);
 
         Ok(Some(all_methods))
     }

